### PR TITLE
feature: SDK Defaults - S3 Params for Session

### PR DESCRIPTION
**Pull request by rubanh**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/s3_utils.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        return ""
+
+    if filtered_args[0].startswith("s3://"):
+        path = str(pathlib.PurePosixPath(*filtered_args[1:])).lstrip("/")
+        return str(pathlib.PurePosixPath(filtered_args[0], path)).replace("s3:/", "s3://")
+
+    return str(pathlib.PurePosixPath(*filtered_args)).lstrip("/")
+
+
+def s3_path_join_with_end_slash(*args):
```
Comment (mufaddal-rohawala) [inline src/sagemaker/s3_utils.py:71]: Not sure why we are marking this as deprecated and then using it as well. Maybe change wording with recommendation.

Couldnt the `s3_path_join` function handle the extra "/" logic edge case.

File: src/sagemaker/config/config_schema.py [hunk 2]
```diff
@@ -477,6 +491,33 @@ def _simple_path(*args: str):
...
+                                            ),
+                                            TYPE: "string",
+                                            # Regex based on
+                                            # https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+                                            # For now, the regex only allows the "safe characters"
+                                            # specified by S3. The regex can be loosened (but not
+                                            # tightened) in the future without introducing
+                                            # backward incompatibility.
+                                            "pattern": (
+                                                r"^[a-zA-Z0-9!_.*'()-]+(/[a-zA-Z0-9!_.*'()-]+)*$"
```
Comment (mufaddal-rohawala) [inline src/sagemaker/config/config_schema.py:514]: is there a reason we start with this pattern (is this based on aws docs)? Can we allow all?

File: src/sagemaker/s3_utils.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    """
+    path = s3_path_join(*args)
+
+    if path:
+        return path + "/"
+
+    return path
+
+
+def calculate_bucket_and_prefix(
```
Comment (mufaddal-rohawala) [inline src/sagemaker/s3_utils.py:86]: nit: can we use naming as determine_bucket_and_prefix/derive_bucket_and_prefix ?

File: src/sagemaker/lambda_helper.py [hunk 4]
```diff
@@ -122,12 +124,13 @@ def create(self):
         if self.script is not None:
             code = {"ZipFile": _zip_lambda_code(self.script)}
         else:
-            bucket = self.s3_bucket or self.session.default_bucket()
+            bucket, key_prefix = s3.calculate_bucket_and_prefix(self.s3_bucket, None, self.session)
```
Comment (mufaddal-rohawala) [inline src/sagemaker/lambda_helper.py:127]: nit: if possible lets use named args for debugging help.

File: src/sagemaker/lambda_helper.py [hunk 5]
```diff
@@ -166,7 +169,10 @@ def update(self):
                         Architectures=self.architectures,
                     )
                 else:
-                    bucket = self.s3_bucket or self.session.default_bucket()
+                    bucket, key_prefix = s3.calculate_bucket_and_prefix(
+                        self.s3_bucket, None, self.session
```
Comment (mufaddal-rohawala) [inline src/sagemaker/lambda_helper.py:173]: nit: if possible lets use named args for debugging help.

File: src/sagemaker/session.py [hunk 6]
```diff
@@ -280,6 +298,19 @@ def _initialize(
             # create a default S3 resource, but only if it needs to fetch from S3
             self.sagemaker_config = load_sagemaker_config(s3_resource=self.s3_resource)
 
+        # after sagemaker_config initialization, update self._default_bucket_name_override if needed
+        self._default_bucket_name_override = resolve_value_from_config(
```
Comment (mufaddal-rohawala) [inline src/sagemaker/session.py:302]: nit: can we rename this to `resolve_value_from_defaults_config`

File: src/sagemaker/workflow/airflow.py [hunk 7]
```diff
@@ -41,8 +41,18 @@ def prepare_framework(estimator, s3_operations):
     elif estimator.uploaded_code is not None:
         bucket, key = s3.parse_s3_url(estimator.uploaded_code.s3_prefix)
     else:
+
+        # estimator.sagemaker_session._default_bucket may be None, unlike
+        # estimator.sagemaker_session.default_bucket.
+        # TODO: should this be updated to use default_bucket instead of _default_bucket? If yes,
```
Comment (mufaddal-rohawala) [inline src/sagemaker/workflow/airflow.py:47]: I dont see why not, we can try and let the integs test it. Nice find!

File: tests/unit/__init__.py [hunk 8]
```diff
@@ -277,3 +299,63 @@
         },
     },
 }
+
+
+def _test_default_bucket_and_prefix_combinations(
```
Comment (mufaddal-rohawala) [inline tests/unit/__init__.py:304]: nice!

File: tests/unit/sagemaker/model/test_deploy.py [hunk 9]
```diff
@@ -67,11 +67,15 @@
 
 SHAP_BASELINE = '1,2,3,"good product"'
 CSV_MIME_TYPE = "text/csv"
+BUCKET_NAME = "mybucket"
 
 
 @pytest.fixture
 def sagemaker_session():
-    session = Mock()
+    session = Mock(
```
Comment (mufaddal-rohawala) [inline tests/unit/sagemaker/model/test_deploy.py:75]: nit: can probably get this fixture in common place like conftest to be reused accross and not requiring changes in multiple places?

File: tests/unit/test_pca.py [hunk 10]
```diff
@@ -54,6 +54,7 @@ def sagemaker_session():
         s3_client=None,
         s3_resource=None,
         settings=SessionSettings(),
+        default_bucket_prefix=None,
```
Comment (mufaddal-rohawala) [inline tests/unit/test_pca.py:57]: nit: can also add this ficture in some common place?

**Review by mufaddal-rohawala — CHANGES_REQUESTED** [review]

File: src/sagemaker/workflow/airflow.py (hunk 7 above)
Comment (rubanh) [reply src/sagemaker/workflow/airflow.py:47]: cool, will try it out!

File: tests/unit/sagemaker/model/test_deploy.py (hunk 9 above)
Comment (rubanh) [reply tests/unit/sagemaker/model/test_deploy.py:75]: Yeah agreed, I wanted to also, will come back to these if I have time

File: src/sagemaker/session.py (hunk 6 above)
Comment (rubanh) [reply src/sagemaker/session.py:302]: sure thing

File: src/sagemaker/lambda_helper.py (hunk 5 above)
Comment (rubanh) [reply src/sagemaker/lambda_helper.py:173]: makes sense

File: src/sagemaker/s3_utils.py (hunk 3 above)
Comment (rubanh) [reply src/sagemaker/s3_utils.py:86]: these are much better names! thanks!

File: src/sagemaker/s3_utils.py (hunk 1 above)
Comment (rubanh) [reply src/sagemaker/s3_utils.py:71]: Ack. I made it for backwards compatibility reasons but wanted to discourage people from using it when adding new code (which is also why I didn't add a param to s3_path_join itself). Will see if I can make clearer

File: src/sagemaker/config/config_schema.py (hunk 2 above)
Comment (rubanh) [reply src/sagemaker/config/config_schema.py:514]: Will think about allowing all. I didnt realize earlier that even something like ```//ghh/./gkjhkjhkj///hf/``` can be a valid prefix. I think the way the SDK currently generates S3 paths could break with a prefix like that though (will try it out)

Comment (balajisankar15) [reply src/sagemaker/config/config_schema.py:514]: I'm fine with this. If we allow "all" now and later if we want to add a regex => then it is not a backwards compatible change.

File: src/sagemaker/config/config_schema.py [hunk 11]
```diff
@@ -89,13 +89,17 @@
 OBJECT = "object"
 ADDITIONAL_PROPERTIES = "additionalProperties"
 ENABLE_INTER_CONTAINER_TRAFFIC_ENCRYPTION = "EnableInterContainerTrafficEncryption"
+SESSION = "Session"
```
Comment (prateekmehrotra) [inline src/sagemaker/config/config_schema.py:92]: This is a little confusing with sagemaker session, what will this be used for.? trying to understand if there is a better name for this.

File: src/sagemaker/amazon/amazon_estimator.py [hunk 12]
```diff
@@ -93,8 +93,14 @@ def __init__(
             enable_network_isolation=enable_network_isolation,
             **kwargs
         )
-        data_location = data_location or "s3://{}/sagemaker-record-sets/".format(
-            self.sagemaker_session.default_bucket()
+
```
Comment (prateekmehrotra) [inline src/sagemaker/amazon/amazon_estimator.py:96]: Did you double check if the old data location also have a "/" in the end.?

File: src/sagemaker/amazon/amazon_estimator.py [hunk 13]
```diff
@@ -93,8 +93,14 @@ def __init__(
...
+        data_location = data_location or (
+            s3_utils.s3_path_join_with_end_slash(
+                "s3://",
+                self.sagemaker_session.default_bucket(),
+                self.sagemaker_session.default_bucket_prefix,
```
Comment (prateekmehrotra) [inline src/sagemaker/amazon/amazon_estimator.py:101]: Is this backward compatible where earlier the prefix was sagemaker-record-sets and now there is one more layer in between.?

File: src/sagemaker/config/__init__.py [hunk 14]
```diff
@@ -93,6 +93,8 @@
     AUTO_ML_VOLUME_KMS_KEY_ID_PATH,
     AUTO_ML_INTER_CONTAINER_ENCRYPTION_PATH,
     ENDPOINT_CONFIG_DATA_CAPTURE_KMS_KEY_ID_PATH,
+    SESSION_S3_BUCKET_PATH,
```
Comment (prateekmehrotra) [inline src/sagemaker/config/__init__.py:96]: Can we quickly discuss about this naming, want to make sure we are using session correctly.

File: src/sagemaker/config/config_schema.py [hunk 15]
```diff
@@ -298,6 +298,11 @@ def _simple_path(*args: str):
 TRAINING_JOB_INTER_CONTAINER_ENCRYPTION_PATH = _simple_path(
     SAGEMAKER, TRAINING_JOB, ENABLE_INTER_CONTAINER_TRAFFIC_ENCRYPTION
 )
+SESSION_S3_BUCKET_PATH = _simple_path(SAGEMAKER, PYTHON_SDK, MODULES, SESSION, S3_BUCKET)
+SESSION_S3_OBJECT_KEY_PREFIX_PATH = _simple_path(
```
Comment (prateekmehrotra) [inline src/sagemaker/config/config_schema.py:302]: We should add doc-string's on what these mean

File: src/sagemaker/config/config_schema.py (hunk 2 above)
Comment (prateekmehrotra) [inline src/sagemaker/config/config_schema.py:514]: I couldn't find this pattern in the public docs, lets check up on this offline

File: src/sagemaker/amazon/amazon_estimator.py (hunk 12 above)
Comment (rubanh) [reply src/sagemaker/amazon/amazon_estimator.py:96]: yep before in the case where the user didnt provide a data_location themselves, there would be a "/" at the end. lmk if Im misunderstanding the question

File: src/sagemaker/amazon/amazon_estimator.py (hunk 13 above)
Comment (rubanh) [reply src/sagemaker/amazon/amazon_estimator.py:101]: yep, behavior will stay exactly the same as now unless and until the Session's default_bucket_prefix is set with a non None/empty value (can see s3_path_join unit tests for cases Ive tested)

**Review by knikure — COMMENTED** [review]
/bot run pr

File: src/sagemaker/config/__init__.py (hunk 14 above)
Comment (rubanh) [reply src/sagemaker/config/__init__.py:96]: SESSION_DEFAULT_S3_BUCKET_PATH
SESSION_DEFAULT_S3_OBJECT_KEY_PREFIX_PATH

File: src/sagemaker/config/config_schema.py [hunk 16]
```diff
@@ -477,6 +491,34 @@ def _simple_path(*args: str):
                             TYPE: OBJECT,
                             ADDITIONAL_PROPERTIES: False,
                             PROPERTIES: {
+                                SESSION: {
+                                    TYPE: OBJECT,
+                                    ADDITIONAL_PROPERTIES: False,
+                                    PROPERTIES: {
+                                        S3_BUCKET: {
```
Comment (rubanh) [inline src/sagemaker/config/config_schema.py:498]: SESSION_DEFAULT_S3_BUCKET

File: src/sagemaker/config/config_schema.py [hunk 17]
```diff
@@ -477,6 +491,34 @@ def _simple_path(*args: str):
...
+                                            "description": "sets `default_bucket` of Session",
+                                            "$ref": "#/definitions/s3Bucket",
+                                        },
+                                        S3_OBJECT_KEY_PREFIX: {
```
Comment (rubanh) [inline src/sagemaker/config/config_schema.py:502]: SESSION_DEFAULT_S3_OBJECT_KEY_PREFIX

File: src/sagemaker/config/config_schema.py (hunk 15 above)
Comment (rubanh) [reply src/sagemaker/config/config_schema.py]: (discussed offline -- these are not surfaced to the SDK user so dont need doc strings)

File: src/sagemaker/workflow/airflow.py (hunk 7 above)
Comment (rubanh) [reply src/sagemaker/workflow/airflow.py]: updated and seems to be passing tests

File: src/sagemaker/config/config_schema.py (hunk 2 above)
Comment (rubanh) [reply src/sagemaker/config/config_schema.py]: removed regex and handled the main edge cases I was worried about by refactoring the S3 utils

File: src/sagemaker/session.py (hunk 6 above)
Comment (rubanh) [reply src/sagemaker/session.py:302]: Will tackle later in a separate PR, time permitting

File: tests/unit/sagemaker/model/test_deploy.py (hunk 9 above)
Comment (rubanh) [reply tests/unit/sagemaker/model/test_deploy.py:75]: Will tackle later in a separate PR, time permitting

File: src/sagemaker/config/config_schema.py (hunk 2 above)
Comment (rubanh) [reply src/sagemaker/config/config_schema.py]: removed regex and handled the main edge cases I was worried about by refactoring the S3 utils

File: src/sagemaker/experiments/_helper.py [hunk 18]
```diff
@@ -75,8 +76,17 @@ def upload_artifact(self, file_path):
             raise ValueError(
                 "{} does not exist or is not a file. Please supply a file path.".format(file_path)
             )
-        if not self.artifact_bucket:
-            self.artifact_bucket = self.sagemaker_session.default_bucket()
+
+        # If self.artifact_bucket is falsy, it will be set to sagemaker_session.default_bucket.
```
Comment (prateekmehrotra) [inline src/sagemaker/experiments/_helper.py:80]: nit: typo

**Review by prateekmehrotra — APPROVED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run slow-tests, notebook-tests
